walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,121 @@
+        Collection<TimeRange> allMeetingTimes = 
+            findMeetingTimes(events, allAttendees, meetingDuration);
+
+        if (!allMeetingTimes.isEmpty()) {
+            return allMeetingTimes;
+        }
+
+        // If there is no required attendee, return allMeetingTimes for 
+        // optional attendees.
+        if (requiredAttendees.isEmpty()) {
+            return allMeetingTimes;
+        }
+
+        // If not, return available TimeRange for required attendees.
+        return findMeetingTimes(
+            events, requiredAttendees, meetingDuration);
+    }
+
+    /**
+      * Find all times that the meeting could happen for given attendees. Loop through 
+      * the given events and update the rangeStart pointer, so that it is pointing at the 
+      * start of the current available TimeRange. If the current available TimeRange has 
+      * longer duration than the requested meeting duration, add it to available meetingTimes. 
+      * Then update the pointer to the end of the current event.
+      */
+    private static Collection<TimeRange> findMeetingTimes(
+        Collection<Event> events, Collection<String> attendees, long duration) {
+        Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+
+        // Sort the events by the start time in acsending order.
+        List<Event> eventsSorted = new ArrayList<Event>(events);
+        Collections.sort(eventsSorted, Event.SORT_BY_START_ASCENDING);
+
+        // Candidate for the start of the current available TimeRange.
+        // It will also be pointing at the end of previous event that 
+        // contains any of the give attendees.
+        int rangeStart = DAY_START;
+
+        // Loop through the sorted events and add new TimeRange to 
+        // meetingTimes.
+        for (Event event : eventsSorted) {
+            Collection<String> eventAttendees = event.getAttendees();
+            TimeRange eventWhen = event.getWhen();
+            int eventStart = eventWhen.start();
+            int eventEnd = eventWhen.end();
+
+            // Check if given attendees shows up in the event.
+            if (isAttending(attendees, eventAttendees)) {
+                if (rangeStart >= eventStart) {
+                    // There is overlap or no gap between the events.
+                    if (eventEnd > rangeStart) {
+                        // Update the pointer so that it is pointing at the start of new 
+                        // available TimeRange.
+                        rangeStart = eventEnd;
+                    }
+                } else {
+                    // If they are separate, initialize the current available TimeRange.            
+                    TimeRange newRange = TimeRange.fromStartEnd(
+                            rangeStart, eventStart, false);
+                        if (newRange.duration() >= duration) {
+                            meetingTimes.add(newRange);
+                        }
```

Comment:
This block doesn't seem properly indented.

Reply:
The indentation is updated.